test/AbstractLinkTest.php

```diff
@@ -533,6 +533,58 @@ public function testTransaction()
         }
     }
 
+    public function testTransactionIsReleasedOnCommit(): void
+    {
+        $transaction = $this->link->beginTransaction();
+
+        $this->assertInstanceOf(Transaction::class, $transaction);
+
+        $data = $this->getData()[0];
+
+        $result = $transaction->execute("SELECT * FROM test WHERE domain=:domain", ['domain' => $data[0]]);
+        unset($result);
+        $transaction->commit();
+
+        try {
+            $transaction->execute("SELECT * FROM test");
+            $this->fail('Query should fail after transaction commit');
+        } catch (TransactionError $exception) {
+            // Exception expected.
+        }
+
+        $this->assertTrue($this->link->isAlive());
+
+        $result = $this->link->execute('SELECT * FROM test WHERE domain = ?', [$data[0]]);
+        self::assertSame(1, $result->getRowCount());
+        unset($result);
+    }
+
+    public function testTransactionIsReleasedOnRollback(): void
+    {
+        $transaction = $this->link->beginTransaction();
+
+        $this->assertInstanceOf(Transaction::class, $transaction);
+
+        $data = $this->getData()[0];
+
+        $result = $transaction->query("DELETE FROM test");
+        unset($result);
+        $transaction->rollback();
+
+        try {
+            $transaction->execute("SELECT * FROM test");
+            $this->fail('Query should fail after transaction commit');
+        } catch (TransactionError $exception) {
+            // Exception expected.
+        }
+
+        $this->assertTrue($this->link->isAlive());
+
+        $result = $this->link->execute('SELECT * FROM test WHERE domain = ?', [$data[0]]);
+        self::assertSame(1, $result->getRowCount());
+        unset($result);
```

Comment:
this still feels buggy to me, even tho it's consistent with v1.

If we remove unset($result) in line 571, we still get the same error, as in line 583, the query is executed before `$result` is destructed, so the transaction is not yet released.

wdyt @trowski?